indexing-service/src/main/java/org/apache/druid/indexing/common/task/AbstractBatchIndexTask.java

```diff
@@ -135,6 +148,49 @@ public void stopGracefully(TaskConfig taskConfig)
     }
   }
 
+  public static FilteringCloseableInputRowIterator inputSourceReader(
+      File tmpDir,
+      DataSchema dataSchema,
+      InputSource inputSource,
+      @Nullable InputFormat inputFormat,
+      Predicate<InputRow> rowFilter,
+      RowIngestionMeters ingestionMeters,
+      ParseExceptionHandler parseExceptionHandler
+  ) throws IOException
+  {
+    final List<String> metricsNames = Arrays.stream(dataSchema.getAggregators())
+                                            .map(AggregatorFactory::getName)
+                                            .collect(Collectors.toList());
+    final InputSourceReader inputSourceReader = dataSchema.getTransformSpec().decorate(
+        inputSource.reader(
+            new InputRowSchema(
+                dataSchema.getTimestampSpec(),
+                dataSchema.getDimensionsSpec(),
+                metricsNames
+            ),
```

Comment:
This is calculated in 3 places in the code - `AbstractBatchIndexTask`, `InputSourceSampler` and `SeekableStreamIndexTaskRunner`. I think it would be a good idea to consolidate these methods, so that metricNames is always calculated the same way given a dataSchema

Reply:
Yeah, it would be useful. Or we can change `InputRowSchema` to have all type information of metrics as well, so that we can check exceptions in parsing and type casting in only `InputEntityReader`. It means, after this change, you will not have to worry about parseExceptions in `IncrementalIndex` anymore. Probably we don't need `ParseExceptionHandler` at all in that case. I haven't done this refactoring here because it will be quite complicated and this PR is already big enough.